framework/util/spirv_parsing_util.cpp

```diff
@@ -0,0 +1,476 @@
+        }
+    }
+    return nullptr;
+}
+
+bool SpirVParsingUtil::GetVariableDecorations(const Instruction*   variable_insn,
+                                              BufferReferenceInfo& buffer_reference_info)
+{
+    const uint32_t variable_id   = variable_insn->resultId();
+    const uint32_t storage_class = variable_insn->operand(0);
+
+    if (storage_class == spv::StorageClassPushConstant)
+    {
+        buffer_reference_info.push_constant_block = true;
+        return true;
+    }
+
+    for (const Instruction* insn : decorations_instructions_)
+    {
+        if (insn->operand(0) != variable_id)
+        {
+            continue;
+        }
+        if (insn->operand(1) == spv::DecorationDescriptorSet)
+        {
+            buffer_reference_info.set = insn->operand(2);
+        }
+        else if (insn->operand(1) == spv::DecorationBinding)
+        {
+            buffer_reference_info.binding = insn->operand(2);
+        }
+    }
+
+    if (storage_class == spv::StorageClassStorageBuffer || storage_class == spv::StorageClassUniform)
+    {
+        return true;
+    }
+    else
+    {
+        printf("Storage class %u not handled\n", storage_class);
+        return false;
+    }
+}
+
+bool SpirVParsingUtil::Parse(const uint32_t* const spirv_code, size_t spirv_num_bytes)
+{
+    if (spirv_code == nullptr)
+    {
+        return false;
+    }
+
+    definitions_.clear();
+    store_instructions_.clear();
+    decorations_instructions_.clear();
+    buffer_reference_map_.clear();
+
+    // use in combination with spirv-reflect
+    std::optional<SpvReflectShaderModule> spv_shader_module;
+
+    // spirv-header is 5 d-words
+    constexpr uint32_t spirv_header_size = 5;
+    const uint32_t*    spirv_ptr         = spirv_code;
+    const uint32_t*    spirv_end         = spirv_code + (spirv_num_bytes / sizeof(uint32_t));
+
+    // skip header
+    spirv_ptr += spirv_header_size;
+
+    std::vector<Instruction> instructions;
+
+    // First build up instructions object to make it easier to work with the SPIR-V
+    while (spirv_ptr != spirv_end)
+    {
+        Instruction& insn = instructions.emplace_back(spirv_ptr);
+        spirv_ptr += insn.length();
+    }
+    instructions.shrink_to_fit();
+
+    auto track_back_instruction = [this, &spv_shader_module, spirv_code, spirv_num_bytes](
+                                      const Instruction* object_insn) {
+        // keep track of access-chain
+        std::vector<uint32_t> access_indices;
+
+        // We are where a buffer-reference was accessed, now walk back to find where it came from
+        while (object_insn)
+        {
+            switch (object_insn->opcode())
+            {
+                case spv::OpConvertUToPtr:
+                case spv::OpCopyLogical:
+                case spv::OpLoad:
+                    object_insn = FindDef(object_insn->operand(0));
+                    break;
+                case spv::OpAccessChain:
```

Comment:
this extraction of access-chain -> indices 'seems' to work properly. does this look sane to you @spencer-lunarg ?

Reply:
something to be aware of in general is there could be 2 `OpAccessChains` chained together

https://github.com/KhronosGroup/SPIRV-Guide/blob/main/chapters/access_chains.md#example---chained-access-chains